conditions = ['DMSO', 'THAP']
replicates = [1, 2, 3, 4]

# cd /rds/project/rs2099/rds-rs2099-toxgenomics/eisa

rule do_fastqc:
    input:
        fastq = expand("{cond}_RNAseq_rep{rep}.fastq",
                    cond=conditions,
                    rep=replicates)
    output:
        html = expand("{cond}_RNAseq_rep{rep}_fastqc.html",
                    cond=conditions,
                    rep=replicates)
    shell:
        """
        fastqc {input.fastq}
        """

rule star_map:
    input:
        fa = "././shared/genome/GRCh38.primary_assembly.genome.fa"
        fastq = expand("{cond}_RNAseq_rep{rep}.fastq",
                    cond=conditions,
                    rep=replicates)
    output:

    shell:
        """STAR --runThreadN 8 \
        --genomeDir {input.fa} \
        --readFilesIn {input.fastq} \
        --outSAMtype BAM SortedByCoordinate \
        --outSAMunmapped Within \
        --outSAMattributes NH HI AS nM jM jI
        """

# rule deseq2_normalise:
# rule annotate:
# count introns and exons
# eisaR
